Mathlib/Data/Matroid/Rank/Finite.lean

```diff
@@ -0,0 +1,190 @@
+  h.finite_of_basis hI
+
+lemma Basis'.rkFin_of_finite (hI : M.Basis' I X) (hIfin : I.Finite) : M.RkFin X :=
+  ⟨I, hI, hIfin⟩
+
+lemma Basis.rkFin_of_finite (hI : M.Basis I X) (hIfin : I.Finite) : M.RkFin X :=
+  ⟨I, hI.basis', hIfin⟩
+
+lemma Basis'.finite_iff_rkFin (hI : M.Basis' I X) : I.Finite ↔ M.RkFin X :=
+  ⟨hI.rkFin_of_finite, fun h ↦ h.finite_of_basis' hI⟩
+
+lemma Basis.finite_iff_rkFin (hI : M.Basis I X) : I.Finite ↔ M.RkFin X :=
+  hI.basis'.finite_iff_rkFin
+
+/-- A `RkFin` set has a finite `Basis'`-/
+lemma RkFin.exists_finite_basis' (h : M.RkFin X) : ∃ I, M.Basis' I X ∧ I.Finite :=
+  h.exists_finite_base
+
+/-- A `RkFin` set has a finset `Basis'` -/
+lemma RkFin.exists_finset_basis' (h : M.RkFin X) : ∃ (I : Finset α), M.Basis' I X :=
+  let ⟨I, hI, hIfin⟩ := h.exists_finite_basis'
+  ⟨hIfin.toFinset, by simpa⟩
+
+/-- A set is `RkFin` iff it has a finite `Basis'` -/
+lemma rkFin_iff_exists_basis' : M.RkFin X ↔ ∃ I, M.Basis' I X ∧ I.Finite :=
+  ⟨RkFin.exists_finite_basis', fun ⟨_, hIX, hI⟩ ↦ hIX.rkFin_of_finite hI⟩
+
+lemma RkFin.subset (h : M.RkFin X) (hXY : Y ⊆ X) : M.RkFin Y := by
+  obtain ⟨I, hI⟩ := M.exists_basis' Y
+  obtain ⟨J, hJ, hIJ⟩ := hI.indep.subset_basis'_of_subset (hI.subset.trans hXY)
+  exact hI.rkFin_of_finite <| (hJ.finite_of_rkFin h).subset hIJ
+
+@[simp]
+lemma rkFin_inter_ground_iff : M.RkFin (X ∩ M.E) ↔ M.RkFin X :=
+  let ⟨_I, hI⟩ := M.exists_basis' X
+  ⟨fun h ↦ hI.rkFin_of_finite (hI.basis_inter_ground.finite_of_rkFin h),
+    fun h ↦ h.subset inter_subset_left⟩
+
+lemma RkFin.to_inter_ground (h : M.RkFin X) : M.RkFin (X ∩ M.E) :=
+  rkFin_inter_ground_iff.2 h
+
+lemma rkFin_iff (hX : X ⊆ M.E := by aesop_mat) : M.RkFin X ↔ ∃ I, M.Basis I X ∧ I.Finite := by
+  simp_rw [rkFin_iff_exists_basis', M.basis'_iff_basis hX]
+
+lemma Indep.rkFin_iff_finite (hI : M.Indep I) : M.RkFin I ↔ I.Finite :=
+  hI.basis_self.finite_iff_rkFin.symm
+
+lemma Indep.finite_of_rkFin (hI : M.Indep I) (hfin : M.RkFin I) : I.Finite :=
+  hI.basis_self.finite_of_rkFin hfin
+
+lemma rkFin_of_finite (M : Matroid α) (hX : X.Finite) : M.RkFin X :=
+  let ⟨_, hI⟩ := M.exists_basis' X
+  hI.rkFin_of_finite (hX.subset hI.subset)
+
+lemma Indep.subset_finite_basis'_of_subset_of_rkFin (hI : M.Indep I) (hIX : I ⊆ X)
+    (hX : M.RkFin X) : ∃ J, M.Basis' J X ∧ I ⊆ J ∧ J.Finite :=
+  (hI.subset_basis'_of_subset hIX).imp fun _ hJ => ⟨hJ.1, hJ.2, hJ.1.finite_of_rkFin hX⟩
+
+lemma Indep.subset_finite_basis_of_subset_of_rkFin (hI : M.Indep I) (hIX : I ⊆ X)
+    (hX : M.RkFin X) (hXE : X ⊆ M.E := by aesop_mat) : ∃ J, M.Basis J X ∧ I ⊆ J ∧ J.Finite :=
+  (hI.subset_basis_of_subset hIX).imp fun _ hJ => ⟨hJ.1, hJ.2, hJ.1.finite_of_rkFin hX⟩
+
+lemma rkFin_singleton (M : Matroid α) (e : α) : M.RkFin {e} :=
+  rkFin_of_finite M (finite_singleton e)
+
+@[simp]
+lemma RkFin.empty (M : Matroid α) : M.RkFin ∅ :=
+  rkFin_of_finite M finite_empty
+
+lemma not_rkFin_superset (h : ¬M.RkFin X) (hXY : X ⊆ Y) : ¬M.RkFin Y :=
+  fun h' ↦ h (h'.subset hXY)
```

Comment:
I'd say a lemma like this doesn't need to exist, since it's just `mt`  with `RkFin.subset` and `hXY`.